build/stm32xx-sys/src/lib.rs

```diff
@@ -0,0 +1,152 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use quote::{quote, TokenStreamExt};
+use serde::Deserialize;
+use std::collections::BTreeMap;
+
+/// This represents our _subset_ of global config and _must not_ be marked with
+/// `deny_unknown_fields`!
+#[derive(Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct GlobalConfig {
+    sys: SysConfig,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "kebab-case", deny_unknown_fields)]
+pub struct SysConfig {
+    /// EXTI interrupts
+    gpio_irqs: BTreeMap<String, GpioIrqConfig>,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "kebab-case", deny_unknown_fields)]
+struct GpioIrqConfig {
+    port: ConfigPort,
+    pin: u8,
+    owner: GpioIrqOwner,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "kebab-case", deny_unknown_fields)]
+struct GpioIrqOwner {
+    name: String,
+    notification: String,
+}
+
+#[derive(Copy, Clone, Debug, Deserialize)]
+enum ConfigPort {
+    A,
+    B,
+    C,
+    D,
+    E,
+    F,
+    G,
+    H,
+    I,
+    J,
+    K,
+}
+
+impl quote::ToTokens for ConfigPort {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        use proc_macro2::{Ident, Punct, Spacing, Span};
+        // i don't like typing
+        macro_rules! table {
+            ($($variant:ident),*) => {
+                match self {
+                    $(ConfigPort::$variant => {
+                        tokens.append(Ident::new("Port", Span::call_site()));
+                        tokens.append(Punct::new(':', Spacing::Joint));
+                        tokens.append(Punct::new(':', Spacing::Alone));
+                        tokens.append(Ident::new(stringify!($variant), Span::call_site()));
+                    }),*
+                }
+            }
+        }
+        table! { A, B, C, D, E, F, G, H, I, J, K }
```

Comment:
somehow this macro felt like "less work" than just hand-writing the `match`...i don't know what's wrong with me